File: model.requisiti/test/it/unibo/SimplePayloadExtractor.java
package it.unibo;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class SimplePayloadExtractor {
    private String id;

    public SimplePayloadExtractor(String id) {
        this.id = id;
    }

    public List<String> extractPayload(String str) {
        Pattern pattern = Pattern.compile(id + "\\(([^\\)]*)\\)");
        Matcher matcher = pattern.matcher(str);
        String payload = null;
        if (matcher.matches()) {
            payload = matcher.group(1);

            if (payload == null) {
                Matcher emptyMatcher = Pattern.compile(id + "\\(\\s*\\)").matcher(str);
                if (!emptyMatcher.matches()) {
                    throw new IllegalArgumentException("Message not properly formatted: expected " + id + "(X, Y...), got " + str);
                }
            }
        } else {
            throw new IllegalArgumentException("Message not properly formatted: expected " + id + "(X, Y...), got " + str);
        }

        List<String> out = Arrays.stream(payload.split("\\s*,\\s*")).collect(Collectors.toList());
        if (out.size() == 1 && out.get(0).equals("")) {
            return new ArrayList<String>();
        } else {
            return out;
        }
    }
}

File: wasteservice/test/it/unibo/lenziguerra/wasteservice/SimplePayloadExtractor.kt
package it.unibo.lenziguerra.wasteservice

import java.util.*
import java.util.regex.Pattern
import java.util.stream.Collectors

class SimplePayloadExtractor(private val id: String) {
    fun extractPayload(str: String): List<String> {
        val pattern = Pattern.compile("$id\\(([^)]*)\\)")
        val matcher = pattern.matcher(str)
        var payload: String? = null
        if (matcher.matches()) {
            payload = matcher.group(1)
            if (payload == null) {
                val emptyMatcher = Pattern.compile("$id\\(\\s*\\)").matcher(str)
                require(emptyMatcher.matches()) { "Message not properly formatted: expected $id(X, Y...), got $str" }
            }
        } else {
            throw IllegalArgumentException("Message not properly formatted: expected $id(X, Y...), got $str")
        }
        val out = Arrays.stream(payload.split("\\s*,\\s*".toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()).collect(Collectors.toList())
        return if (out.size == 1 && out[0] == "") {
            ArrayList()
        } else {
            out
        }
    }
}